cmd/controller/state/delta/controller.go

```diff
@@ -115,18 +163,57 @@ func (c *Controller) sendDeltas(ctx context.Context, firstDeltaReport bool) {
 		_ = deltaStream.CloseSend()
 	}()
 
-	var sendErr error
+	var sentDeltasCount int
 	for _, item := range pendingDeltas {
 		pbItem := c.toCastaiDelta(item)
-		if err := deltaStream.Send(pbItem); err != nil && !errors.Is(err, io.EOF) {
-			sendErr = err
+		if err := c.sendDeltaItem(ctx, deltaStream, pbItem); err != nil {
+			c.log.Warnf("sending delta item: %v", err)
+			// Return any remaining items back to pending list.
+			c.upsertPendingItems(pendingDeltas[sentDeltasCount:])
+			return
 		}
+		sentDeltasCount++
```

Comment:
Nitpick: `sentDeltasCount` will always be the same as `len(pendingDeltas)`

Reply:
sentDeltasCount is only increased if it was able to send. If send fails loops exists.

Reply:
I guess we should then change line `173` to `break`, as of right now we return.